src/main/java/com/example/processmanagementtool/prospect/ProspectService.java

```diff
@@ -0,0 +1,98 @@
+    private final TemplateRepository templateRepository;
+    private final UserRepository userRepository;
+
+    public Mono<SuccessResponseDTO> createNewProspect(Principal principal, Mono<ProspectRequestDTO> request) {
+        return userRepository.findUserByLogin(principal.getName())
+                .flatMap(foundUser -> findTemplateAndValidateData(request, foundUser.getId())
+                        .flatMap(validatedData -> mapValidatedRequestToProspectAndSave(validatedData, foundUser.getId())))
+                .switchIfEmpty(Mono.error(new UserNotFoundException()));
+    }
+
+    public Mono<ProspectPageDTO> findAllProspectsByUser(Principal principal, PageRequest pageRequest) {
+        return userRepository.findUserByLogin(principal.getName())
+                .flatMap(foundUser -> getMapAndZipProspects(foundUser, pageRequest))
+                .map(res -> new PageImpl<>(res.getT1(), pageRequest, res.getT2()))
+                .flatMap(this::mapProspectPageToProspectPageDTO);
+    }
+
+    private Mono<ProspectPageDTO> mapProspectPageToProspectPageDTO(Page<Prospect> page) {
+        return Mono.just(ProspectPageDTO.builder()
+                .totalPages(page.getTotalPages())
+                .totalElements(page.getTotalElements())
+                .pageElements(page.getNumberOfElements())
+                .content(ProspectDTOMapper.mapProspectToResponse(page.getContent()))
+                .responseType("template")
+                .build());
+    }
+
+    private Mono<Tuple2<List<Prospect>, Long>> getMapAndZipProspects(User user, PageRequest pageRequest) {
+        return prospectRepository.findAllByUserId(user.getId(), pageRequest).collectList()
+                .zipWith(templateRepository.findAllByUserId(user.getId()).count());
+    }
+
+    private Mono<ProspectRequestDTO> findTemplateAndValidateData(Mono<ProspectRequestDTO> request, Long userId) {
+        return request.flatMap(data -> templateRepository.findTemplateByUserIdAndId(userId, data.getTemplateId())
+                        .flatMap(foundTemplate -> validateProspectProperties(foundTemplate, data)))
+                .switchIfEmpty(Mono.error(new BadRequest("Template not found")));
+    }
+
+    private Mono<ProspectRequestDTO> validateProspectProperties(Template template, ProspectRequestDTO request) {
+        List<String> wrongKeys = template.getFields().entrySet()
+                .stream()
+                .filter(entry -> entry.getValue().getRequired())
+                .filter(property -> checkIfPropertyIsNull(property.getKey(), request.getFields()))
+                .map(Map.Entry::getKey)
+                .toList();
+
+        if (!wrongKeys.isEmpty()) {
+            return Mono.error(new BadRequest("Required fields are not filled"));
+        }
+        return Mono.just(request);
+    }
+
+    private boolean checkIfPropertyIsNull(String key, Map<String, ProspectPropertyDTO> requestFields) {
+        ProspectPropertyDTO value = requestFields.get(key);
```

Comment:
moze byc to w 1 lini a nawet w filtrze, i tyle:

Objects.isNull(requestFields.get(key).getData());